File: app/src/main/java/nl/jpelgrm/movienotifier/data/APIClient.java
package nl.jpelgrm.movienotifier.data;

import java.util.List;

import nl.jpelgrm.movienotifier.models.NotificationType;
import nl.jpelgrm.movienotifier.models.User;
import nl.jpelgrm.movienotifier.models.UserLogin;
import nl.jpelgrm.movienotifier.models.Watcher;
import okhttp3.ResponseBody;
import retrofit2.Call;
import retrofit2.http.Body;
import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.Header;
import retrofit2.http.POST;
import retrofit2.http.PUT;
import retrofit2.http.Path;

public interface APIClient {
    // User
    @PUT("/user")
    Call<User> addUser(@Body User user);

    @POST("/login")
    Call<User> login(@Body UserLogin user);

    @GET("/user/{userid}")
    Call<User> getUser(@Header("APIKEY") String apikey, @Path("userid") String userid);

    @POST("/user/{userid}")
    Call<User> updateUser(@Header("APIKEY") String apikey, @Path("userid") String userid, @Body User user);

    @DELETE("/user/{userid}")
    Call<ResponseBody> deleteUser(@Header("APIKEY") String apikey, @Path("userid") String userid);


    // Watcher
    @GET("/watchers")
    Call<List<Watcher>> getWatchers(@Header("APIKEY") String apikey);

    @PUT("/watchers")
    Call<Watcher> addWatcher(@Header("APIKEY") String apikey, @Body Watcher watcher);

    @GET("/watchers/{watcherid}")
    Call<Watcher> getWatcher(@Header("APIKEY") String apikey, @Path("watcherid") String watcherid);

    @POST("/watchers/{watcherid}")
    Call<Watcher> updateWatcher(@Header("APIKEY") String apikey, @Path("watcherid") String watcherid, @Body Watcher watcher);

    @DELETE("/watchers/{watcherid}")
    Call<ResponseBody> deleteWatcher(@Header("APIKEY") String apikey, @Path("watcherid") String watcherid);


    // Notification type
    @GET("/notificationtypes")
    Call<List<NotificationType>> getNotificationTypes(@Header("APIKEY") String apikey);

    @GET("/notificationtype/{notificationtypekey}")
    Call<NotificationType> getNotificationType(@Header("APIKEY") String apikey, @Path("notificationtypekey") String notificationtypekey);
}

File: app/src/main/java/nl/jpelgrm/movienotifier/data/APIClient.kt
package nl.jpelgrm.movienotifier.data

import nl.jpelgrm.movienotifier.models.Cinema
import nl.jpelgrm.movienotifier.models.User
import nl.jpelgrm.movienotifier.models.UserLogin
import nl.jpelgrm.movienotifier.models.Watcher
import okhttp3.ResponseBody
import retrofit2.Call
import retrofit2.http.*

interface APIClient {
    // User
    @PUT("/user")
    fun addUser(@Body user: User): Call<User>

    @POST("/login")
    fun login(@Body user: UserLogin): Call<User>

    @GET("/user/{userid}")
    fun getUser(@Header("APIKEY") apikey: String, @Path("userid") userid: String): Call<User>

    @POST("/user/{userid}")
    fun updateUser(@Header("APIKEY") apikey: String, @Path("userid") userid: String, @Body user: User): Call<User>

    @DELETE("/user/{userid}")
    fun deleteUser(@Header("APIKEY") apikey: String, @Path("userid") userid: String): Call<ResponseBody>

    // Watcher
    @GET("/watchers")
    fun getWatchers(@Header("APIKEY") apikey: String): Call<List<Watcher>>

    @PUT("/watchers")
    fun addWatcher(@Header("APIKEY") apikey: String, @Body watcher: Watcher): Call<Watcher>

    @GET("/watchers/{watcherid}")
    fun getWatcher(@Header("APIKEY") apikey: String, @Path("watcherid") watcherid: String): Call<Watcher>

    @POST("/watchers/{watcherid}")
    fun updateWatcher(@Header("APIKEY") apikey: String, @Path("watcherid") watcherid: String, @Body watcher: Watcher): Call<Watcher>

    @DELETE("/watchers/{watcherid}")
    fun deleteWatcher(@Header("APIKEY") apikey: String, @Path("watcherid") watcherid: String): Call<ResponseBody>

    // Cinemas
    @GET("/cinemas")
    suspend fun getCinemas(): List<Cinema>
}